File: problem-solving/solutions/src/main/java/xyz/grind/coding/blind75/ValidAnagram.java
package xyz.grind.coding.blind75;

import java.util.Arrays;
import java.util.HashSet;

public class ValidAnagram {

   public static boolean isAnagramSortingBruteForce(String s, String t){
      if(s.length() != t.length()) return false;
      char[] sCharArray = s.toCharArray();
      char[] tCharArray = t.toCharArray();

      Arrays.sort(sCharArray);
      Arrays.sort(tCharArray);

      return Arrays.equals(sCharArray, tCharArray);
   }

   public static boolean isAnagramHashSet(String s, String t) {
      if(s.length() != t.length()) return false;

      char[] sCharArray = s.toCharArray();
      char[] tCharArray = t.toCharArray();
      HashSet<Character> set = new HashSet<>();

      for(char sChar : sCharArray) set.add(sChar);
      for(char tChar : tCharArray) {
         if(set.add(tChar)) return false;
      }
      return true;
   }

   public static boolean isAnagram(String s, String t) {
      if(s.length() != t.length()) return false;

      int[] counts = new int[26];

      for(int i = 0; i < s.length(); i++){
         counts[s.charAt(i) - 'a']++;
         counts[t.charAt(i) - 'a']--;
      }

      for(int count : counts) {
         if(count != 0) return false;
      }
      return true;
   }

}

File: problem-solving/solutions/leetcode/blind75/src/main/kotlin/xyz/grind/coding/easy/ValidAnagram.kt
package xyz.grind.coding.easy

import java.util.Arrays

object ValidAnagram {
    fun isAnagramSortingBruteForce(s: String, t: String): Boolean {
        if (s.length != t.length) return false
        val sCharArray = s.toCharArray()
        val tCharArray = t.toCharArray()
        Arrays.sort(sCharArray)
        Arrays.sort(tCharArray)
        return sCharArray.contentEquals(tCharArray)
    }

    fun isAnagramHashSet(s: String, t: String): Boolean {
        if (s.length != t.length) return false
        val sCharArray = s.toCharArray()
        val tCharArray = t.toCharArray()
        val set = hashSetOf<Char>()
        for (sChar in sCharArray) set.add(sChar)
        for (tChar in tCharArray) {
            if (set.add(tChar)) return false
        }
        return true
    }

        fun isAnagramAdjusted(s: String, t: String): Boolean {
            if (s.length != t.length) return false
            val counts = IntArray(26)
            for (i in s.indices) {
                counts[s[i].code - 'a'.code]++
                counts[t[i].code - 'a'.code]--
            }
            for (count in counts) {
                if (count != 0) return false
            }
            return true
        }

    fun isAnagram(s: String, t: String): Boolean {
        if (s.length != t.length) return false
        val counts = IntArray(26)
        for(char in s.indices) {
            counts[s[char] - 'a']++
            counts[t[char] - 'a']--
        }
        return counts.all { it == 0 }
    }
}
